### DISCUSSION TITLE
Implement a way to discover functions by parts of their signatures (a Hoogle equivalent)

### DISCUSSION THREAD

**Discussion by cognivore** [gleam-lang/gleam#1340] [Ideas & suggestions]
# Why?

Types allow us to write a lot of code mechanically, if there is a way to discover functions that return the types we want. Currently, the best way to do that is to check out the entirety of gleam universe and ripgrep it. As gleam ecosystem grows, that shall become not feasible for a regular user.

# What?

 - Pick one or both
  - [ ] Provide a frontend for ripgrepping a la hackage-search.serokell.io
  - [ ] Come up with a data structure and query mechanics for gleam types a la https://hoogle.haskell.org/

# How?

TBD

**Comment by cognivore** [comment]
I think that the first tool to be implemented should be the proper hoogle equivalent.
When I write Gleam I agressively omit type annotations where possible. If others do the same, we will need to be able to search inferred types.

**Comment by lpil** [comment]
A fantastic suggestion, I would love to see a tool like this for Gleam.

One thing to consider is how to we identify Gleam packages on Hex. Perhaps to start it could be a fixed list of package, but later it would be good to automatically add packages as they are uploaded.

> When I write Gleam I agressively omit type annotations where possible. If others do the same, we will need to be able to search inferred types.

One good thing about Gleam is that no type annotations are required for full type inference. It is recommended that you use annotations, but even if you do not we can still extract all the type information for this tool.

**Comment by cognivore** [comment]
FTR, I don't write annotations mostly to stress-test inference.
Normally I write annotations, especially where I am not sure what I'm doing (which happens a lot).

**Comment by lpil** [comment]
@pd-andy suggests the name "gloogle"

**Reply by inoas** [reply]
I'd suggest: `gloggles` - to make sure there is no sueing :D
